elk/training/ccs_reporter.py

```diff
@@ -0,0 +1,318 @@
+            theta /= theta.norm()
+            probe.weight.data = theta[:, :-1]
+            probe.bias.data = theta[:, -1]
+
+        elif self.config.init == "default":
+            for layer in self.probe:
+                if isinstance(layer, nn.Linear):
+                    layer.reset_parameters()
+
+        elif self.config.init == "zero":
+            for param in self.parameters():
+                param.data.zero_()
+        elif self.config.init != "pca":
+            raise ValueError(f"Unknown init: {self.config.init}")
+
+    def forward(self, x: Tensor) -> Tensor:
+        """Return the raw score output of the probe on `x`."""
+        return self.probe(x).squeeze(-1)
+
+    def predict(self, x_pos: Tensor, x_neg: Tensor) -> Tensor:
+        return 0.5 * (self(x_pos).sigmoid() + (1 - self(x_neg).sigmoid()))
+
+    def loss(
+        self,
+        logit0: Tensor,
+        logit1: Tensor,
+        labels: Optional[Tensor] = None,
+    ) -> Tensor:
+        """Return the loss of the reporter on the contrast pair (x0, x1).
+
+        Args:
+            logit0: The raw score output of the reporter on x0.
+            logit1: The raw score output of the reporter on x1.
+            labels: The labels of the contrast pair. Defaults to None.
+
+        Returns:
+            loss: The loss of the reporter on the contrast pair (x0, x1).
+
+        Raises:
+            ValueError: If `supervised_weight > 0` but `labels` is None.
+        """
+        loss = self.unsupervised_loss(logit0, logit1)
+
+        # If labels are provided, use them to compute a supervised loss
+        if labels is not None:
+            num_labels = len(labels)
+            assert num_labels <= len(logit0), "Too many labels provided"
+            p0 = logit0[:num_labels].sigmoid()
+            p1 = logit1[:num_labels].sigmoid()
+
+            alpha = self.config.supervised_weight
+            preds = p0.add(1 - p1).mul(0.5).squeeze(-1)
+            bce_loss = bce(preds, labels.type_as(preds))
+            loss = alpha * bce_loss + (1 - alpha) * loss
+
+        elif self.config.supervised_weight > 0:
+            raise ValueError(
+                "Supervised weight > 0 but no labels provided to compute loss"
+            )
+
+        return loss
+
+    def fit(
+        self,
+        x_pos: Tensor,
+        x_neg: Tensor,
+        labels: Optional[Tensor] = None,
+    ) -> float:
+        """Fit the probe to the contrast pair (x0, x1).
+
+        Args:
+            contrast_pair: A tuple of tensors, (x0, x1), where x0 and x1 are the
+                contrastive representations.
+            labels: The labels of the contrast pair. Defaults to None.
+
+        Returns:
+            best_loss: The best loss obtained.
+
+        Raises:
+            ValueError: If `optimizer` is not "adam" or "lbfgs".
+            RuntimeError: If the best loss is not finite.
+        """
+        self.update(x_pos, x_neg)
```

Comment:
This running mean isn't ever used for this class

Reply:
yeah I need to add this, I was intending to fix #96

Reply:
for some reason the idea of implementing this rn is stressing me out so I'm just not gonna